framework/src/org/checkerframework/framework/util/ContractsUtils.java

```diff
@@ -221,19 +223,87 @@ public int hashCode() {
         return result;
     }
 
-    /** Returns the annotation mirror as specified by the "qualifier" value in metaAnno. */
-    private AnnotationMirror getAnnotationMirrorOfQualifier(AnnotationMirror metaAnno) {
+    /**
+     * Returns the annotation mirror as specified by the "qualifier" value in {@code qualifierAnno}.
+     * If {@code argumentAnno} is specified, then arguments are copied from {@code argumentAnno} to
+     * the returned annotation, renamed according to {@code argumentMap}.
+     *
+     * @param qualifierAnno annotation specifying the qualifier class
+     * @param argumentAnno annotation containing the argument values
+     * @param argumentMap map of argument names, which maps from names in {@code argumentAnno} to
+     *     names used in the returned annotation
+     */
+    private AnnotationMirror getAnnotationMirrorOfQualifier(
+            AnnotationMirror qualifierAnno,
+            AnnotationMirror argumentAnno,
+            Map<String, String> argumentMap) {
+
         @SuppressWarnings("unchecked")
         Class<? extends Annotation> c =
                 (Class<? extends Annotation>)
-                        AnnotationUtils.getElementValueClass(metaAnno, "qualifier", false);
-        AnnotationMirror anno = AnnotationBuilder.fromClass(factory.getElementUtils(), c);
+                        AnnotationUtils.getElementValueClass(qualifierAnno, "qualifier", false);
+
+        AnnotationMirror anno;
+        if (argumentAnno == null || argumentMap.isEmpty()) {
+            // If there are no arguments, use factory method that allows caching
+            anno = AnnotationBuilder.fromClass(factory.getElementUtils(), c);
+        } else {
+            AnnotationBuilder builder = new AnnotationBuilder(factory.getProcessingEnv(), c);
+            builder.copyRenameElementValuesFromAnnotation(argumentAnno, argumentMap);
+            anno = builder.build();
+        }
+
         if (factory.isSupportedQualifier(anno)) {
             return anno;
         } else {
-            return null;
+            AnnotationMirror aliasedAnno = factory.aliasedAnnotation(anno);
+            if (factory.isSupportedQualifier(aliasedAnno)) {
+                return aliasedAnno;
+            } else {
+                return null;
+            }
         }
     }
+    /**
+     * Returns the annotation mirror as specified by the "qualifier" value in {@code contractAnno}.
+     */
+    private AnnotationMirror getAnnotationMirrorOfContractAnnotation(
+            AnnotationMirror contractAnno) {
+        return getAnnotationMirrorOfQualifier(contractAnno, null, null);
+    }
+    /** Makes a map from source argument names to target argument names. */
```

Comment:
I think the javadoc on this method should also mention how the lists don't need to be the same length, but do need to be correlated.